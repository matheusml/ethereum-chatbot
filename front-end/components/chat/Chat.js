import React from "react";
import { StyleSheet, View } from "react-native";

import Header from "../header/Header";
import MessageList from "./messages/MessageList";
import MessageInput from "./messages/MessageInput";

export default class Chat extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      messages: []
    };

    this.sendMessage = this.sendMessage.bind(this);
  }

  sendMessage(newMessage) {
    const { messages } = this.state;

    this.setState({
      messages: [...this.state.messages, newMessage]
    });
  }

  render() {
    return (
      <View style={styles.container}>
        <Header title={"CHANNEL"} />
        <MessageList messages={this.state.messages} />
        <MessageInput sendMessage={this.sendMessage} />
      </View>
    );
  }
}

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#fff"
  }
});
