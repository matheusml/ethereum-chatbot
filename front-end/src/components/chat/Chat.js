import React from "react";
import { StyleSheet, View, ActivityIndicator } from "react-native";
import PropTypes from "prop-types";

import Header from "../header/Header";
import MessageList from "./messages/MessageList";
import MessageInput from "./messages/MessageInput";
import * as messagesAPI from "../../services/messages";
import firebaseApp from "../../config/firebase";

export default class Chat extends React.Component {
  constructor(props) {
    super(props);

    this.state = {
      messages: [],
      loading: false
    };

    this.messagesRef = firebaseApp
      .database()
      .ref()
      .child(props.navigation.state.params.channel);

    this.sendMessage = this.sendMessage.bind(this);
  }

  componentDidMount() {
    // listening for changes on firebase real time database
    this.messagesRef.on("value", snap => {
      const messages = [];
      snap.forEach(child => {
        messages.push({
          sender: child.val().sender,
          message: child.val().message,
          _key: child.key
        });
      });

      this.setState({
        messages
      });
    });
  }

  sendMessage(newMessage) {
    this.setState({ loading: true });
    this.messagesRef.push(newMessage);

    messagesAPI
      .sendMessage()
      .then(response => response.json())
      .then(message => {
        this.messagesRef.push(message);
        this.setState({
          loading: false
        });
      })
      .catch(() => {
        this.setState({ loading: false });
      });
  }

  render() {
    const { navigation } = this.props;
    return (
      <View style={styles.container}>
        <Header title={navigation.state.params.channel} />
        {this.state.loading && (
          <View style={styles.loading}>
            <ActivityIndicator size="large" color="white" />
          </View>
        )}
        <MessageList messages={this.state.messages} />
        <MessageInput
          loading={this.state.loading}
          sendMessage={this.sendMessage}
        />
      </View>
    );
  }
}

Chat.propTypes = {
  navigation: PropTypes.shape({
    state: PropTypes.shape({
      params: PropTypes.shape({
        channel: PropTypes.string.isRequired
      })
    })
  })
};

const styles = StyleSheet.create({
  container: {
    flex: 1,
    backgroundColor: "#fff"
  },
  loading: {
    zIndex: 1,
    position: "absolute",
    left: 0,
    right: 0,
    top: 0,
    bottom: 0,
    alignItems: "center",
    justifyContent: "center",
    backgroundColor: "black",
    opacity: 0.6
  }
});
