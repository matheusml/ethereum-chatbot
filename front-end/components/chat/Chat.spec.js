import React from "react";

import Chat from "./Chat";

import renderer from "react-test-renderer";

describe("Chat", () => {
  it("should match snapshot", () => {
    const tree = renderer.create(<Chat />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  it("should have an initial state", () => {
    const instance = renderer.create(<Chat />).getInstance();
    expect(instance.state).toEqual({ messages: [] });
  });

  it("should update the state when trigerring sendMessage", () => {
    const instance = renderer.create(<Chat />).getInstance();
    const newMessage = {
      message: "new message",
      sender: "sender"
    };
    instance.sendMessage(newMessage);
    expect(instance.state).toEqual({ messages: [newMessage] });
  });
});
