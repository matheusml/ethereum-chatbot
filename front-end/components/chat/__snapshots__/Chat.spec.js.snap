// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Chat should match snapshot 1`] = `
<View
  style={
    Object {
      "backgroundColor": "#fff",
      "flex": 1,
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "lightseagreen",
        "height": 80,
        "justifyContent": "flex-end",
        "padding": 10,
      }
    }
  >
    <Text
      accessible={true}
      allowFontScaling={true}
      ellipsizeMode="tail"
      style={
        Object {
          "color": "white",
          "fontSize": 24,
          "fontWeight": "bold",
        }
      }
    >
      #
      CHANNEL
    </Text>
  </View>
  <RCTScrollView
    data={Array []}
    disableVirtualization={false}
    getItem={[Function]}
    getItemCount={[Function]}
    horizontal={false}
    initialNumToRender={10}
    inverted={true}
    keyExtractor={[Function]}
    maxToRenderPerBatch={10}
    numColumns={1}
    onContentSizeChange={[Function]}
    onEndReachedThreshold={2}
    onLayout={[Function]}
    onMomentumScrollEnd={[Function]}
    onScroll={[Function]}
    onScrollBeginDrag={[Function]}
    onScrollEndDrag={[Function]}
    renderItem={[Function]}
    scrollEventThrottle={50}
    stickyHeaderIndices={Array []}
    style={
      Array [
        Object {
          "transform": Array [
            Object {
              "scaleY": -1,
            },
          ],
        },
        undefined,
      ]
    }
    updateCellsBatchingPeriod={50}
    viewabilityConfigCallbackPairs={Array []}
    windowSize={21}
  >
    <View />
  </RCTScrollView>
  <View
    keyboardVerticalOffset={0}
    onLayout={[Function]}
    style={
      Array [
        undefined,
        Object {
          "paddingBottom": 0,
        },
      ]
    }
  >
    <View
      style={
        Object {
          "backgroundColor": "#eee",
          "flexDirection": "row",
        }
      }
    >
      <TextInput
        allowFontScaling={true}
        onChangeText={[Function]}
        placeholder="Send a message"
        style={
          Object {
            "flex": 1,
            "fontSize": 18,
            "paddingHorizontal": 20,
          }
        }
        underlineColorAndroid="transparent"
        value=""
      />
      <View
        accessibilityComponentType={undefined}
        accessibilityLabel={undefined}
        accessibilityTraits={undefined}
        accessible={true}
        collapsable={undefined}
        hasTVPreferredFocus={undefined}
        hitSlop={undefined}
        isTVSelectable={true}
        nativeID={undefined}
        onLayout={undefined}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          Object {
            "opacity": 1,
          }
        }
        testID={undefined}
        tvParallaxProperties={undefined}
      >
        <Text
          accessible={true}
          allowFontScaling={true}
          ellipsizeMode="tail"
          style={
            Object {
              "alignSelf": "center",
              "color": "lightseagreen",
              "fontSize": 16,
              "fontWeight": "bold",
              "padding": 20,
            }
          }
        >
          Send
        </Text>
      </View>
    </View>
  </View>
</View>
`;
